### ISSUE TITLE
Error: Result: this device doesn't support rfmon (monitor) mode (-6)

### ISSUE THREAD

**Issue by nittanygeek** [dfrencham/swiftPCap#1]
I'm a bit of a noob at this, but even just using "en0" (which is the Ethernet interface on my MacBook Pro) and the default setup from your code, I'm getting an error when running that the device doesn't support rfmon:

```
$ ./swiftPCap
pcap test
Opening en0
Error: Result: this device doesn't support rfmon (monitor) mode (-6)
Exiting
```

**Comment by dfrencham** [comment]
Are you executing it with elevated privileges ?

```
sudo ./swiftPCap
```

**Comment by nittanygeek** [comment]
Yes, I've tried both and get the same results.

**Comment by dfrencham** [comment]
Are you able to post your ifconfig output? (might be a good idea to remove the IPs before posting).

**Comment by nittanygeek** [comment]
It's currently plugged into a network tap:

```
en0: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
    options=10b<RXCSUM,TXCSUM,VLAN_HWTAGGING,AV>
    ether 3c:07:54:21:89:d2 
    inet6 fe80::3e07:54ff:fe21:89d2%en0 prefixlen 64 scopeid 0x4 
    nd6 options=1<PERFORMNUD>
    media: autoselect (100baseTX <full-duplex,flow-control>)
    status: active
```

**Comment by nittanygeek** [comment]
Note: I can fire up a Wireshark session on the interface and it works.  Decided to learn how to write my own and found your article (great writeup, btw).

**Comment by dfrencham** [comment]
cheers :)

Are you using one of the earlier generation Macbook Pro's with the built in ethernet port? or a USB dongle? (I'm asking so I can check out the network chipset and find out if there are any quirks with it).

**Comment by nittanygeek** [comment]
It's a 2011 MacBook Pro with a built in Ethernet.

**Comment by nittanygeek** [comment]
Stepping a little more though the code and it seems pcap_activate() is returning -6.  Both pcap_set_promisc() and pcap_set_rfmon() returns 0, and it's not until pcap_activate() is called that it's passing -6 as the result to handleResult().  I don't know what -6 means, and the docs don't seem to point to anything regarding this type of return:

https://developer.apple.com/library/mac/documentation/Darwin/Reference/ManPages/man3/pcap_activate.3pcap.html

_Edit: Nvm, I see it now in the pcap.h:_

```
#define PCAP_ERROR_RFMON_NOTSUP     -6  /* this device doesn't support rfmon (monitor) mode */
```

**Comment by bchynds** [comment]
Nvm, I see it now in the pcap.h:
```
#define PCAP_ERROR_RFMON_NOTSUP		-6	/* this device doesn't support rfmon (monitor) mode */
```

**Comment by dfrencham** [comment]
Ok, I've discovered that monitor mode is only for WIFI cards, (as opposed to promiscuous mode).

Try commenting out pcap_set_rfmon() and trying again.

**Comment by dfrencham** [comment]
Also - this page has some very helpful information about pcap: http://www.devdungeon.com/content/using-libpcap-c

**Comment by nittanygeek** [comment]
Haha, I literally just discovered this myself as I was asking on the #macdev channel on freenode.  Once I commented it out, it started capturing.  Thanks so much for the help.  I'll definitely check out that link, too.  I was mostly using `man pcap` and your article to get started.  Great stuff!  Thanks again.

**Comment by nittanygeek** [comment]
Just as a quick follow up, I tried to capture on a live network and it wasn't working initially.  This seems to be because the timeout isn't being set in your code.  However, instead of setting the timeout, trigger immediate mode on like so:
`handleResult(pcap_set_immediate_mode(pcapSession, 1), message: "Couldn't set immediate mode")`
before activating.  Once I did this, the packets started showing up immediately instead of filling up in the kernels buffer.
